### PR TITLE
Problem: x/params is deprecated, need to migrate away from it

### LINKED ISSUE #632: Problem: x/params is deprecated, need to migrate away from it
Migrate modules away from `x/params`.

### PR DESCRIPTION
**Pull request by adu-crypto**
👮🏻👮🏻👮🏻 !!!! REFERENCE THE PROBLEM YOUR ARE SOLVING IN THE PR TITLE AND DESCRIBE YOUR SOLUTION HERE !!!! DO NOT FORGET !!!! 👮🏻👮🏻👮🏻
close #632 

According to the recent [Cosmos-SDK ADR 046: Module Params](https://github.com/cosmos/cosmos-sdk/blob/main/docs/architecture/adr-046-module-params.md) , we should migrate `x/cronos` module away from `x/params` to enable stateful params changes, as well as more efficient params serialization.
- [x] store params in `x/cronos` store instead of `subspace`
- [x] create and handle `MsgUpdateParams` for grpc `msgServer`
- [x] support querying params of `x/cronos`
- [x] write and register migration plan 
- [x] add migration test

# PR Checklist:

- [ ] Have you read the [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md)?
- [ ] Does your PR follow the [C4 patch requirements](https://rfc.zeromq.org/spec:42/C4/#23-patch-requirements)?
- [ ] Have you rebased your work on top of the latest master? 
- [ ] Have you checked your code compiles? (`make`)
- [ ] Have you included tests for any non-trivial functionality?
- [ ] Have you checked your code passes the unit tests? (`make test`)
- [ ] Have you checked your code formatting is correct? (`go fmt`)
- [ ] Have you checked your basic code style is fine? (`golangci-lint run`)
- [ ] If you added any dependencies, have you checked they do not contain any known vulnerabilities? (`go list -json -m all | nancy sleuth`)
- [ ] If your changes affect the client infrastructure, have you run the integration test?
- [ ] If your changes affect public APIs, does your PR follow the [C4 evolution of public contracts](https://rfc.zeromq.org/spec:42/C4/#26-evolution-of-public-contracts)?
- [ ] If your code changes public APIs, have you incremented the crate version numbers and documented your changes in the [CHANGELOG.md](https://github.com/crypto-org-chain/chain-main/blob/master/CHANGELOG.md)?
- [ ] If you are contributing for the first time, please read the agreement in [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md#developer-certificate-of-originn).

Thank you for your code, it's appreciated! :)

### CODE REVIEW

**Comment by thomas-nguy** [comment]
does it needs to be handled in other modules as well ? (ibc-go, gravity-bridge, ethermint)

**Comment by adu-crypto** [comment]
> does it needs to be handled in other modules as well ? (ibc-go, gravity-bridge, ethermint)

I think so as these modules also use the legacy `subspace` to get/set params

File: x/cronos/genesis.go [hunk 1]
```diff
@@ -12,7 +12,10 @@ import (
 // InitGenesis initializes the capability module's state from a provided genesis
 // state.
 func InitGenesis(ctx sdk.Context, k keeper.Keeper, genState types.GenesisState) {
-	k.SetParams(ctx, genState.Params)
+	err := k.SetParams(ctx, genState.Params)
+	if err != nil {
```
Comment (JayT106) [inline x/cronos/genesis.go:16]: nit, if err := k.SetParams(ctx, genState.Params) ; err != nil { ...

File: x/cronos/module.go [hunk 2]
```diff
@@ -150,6 +154,12 @@ func (am AppModule) LegacyQuerierHandler(legacyQuerierCdc *codec.LegacyAmino) sd
 // module-specific GRPC queries.
 func (am AppModule) RegisterServices(cfg module.Configurator) {
 	types.RegisterQueryServer(cfg.QueryServer(), am.keeper)
+
+	migrator := keeper.NewMigrator(am.keeper, am.legacySubspace)
+	err := cfg.RegisterMigration(types.ModuleName, 1, migrator.Migrate1to2)
+	if err != nil {
```
Comment (JayT106) [inline x/cronos/module.go:160]: nit

File: x/cronos/keeper/keeper.go [hunk 3]
```diff
@@ -47,29 +48,24 @@ func NewKeeper(
...
 	evmKeeper types.EvmKeeper,
 	accountKeeper types.AccountKeeper,
+	authority string,
 	// this line is used by starport scaffolding # ibc/keeper/parameter
 ) *Keeper {
-	// set KeyTable if it has not already been set
-	if !paramSpace.HasKeyTable() {
-		paramSpace = paramSpace.WithKeyTable(types.ParamKeyTable())
-	}
-
```
Comment (JayT106) [inline x/cronos/keeper/keeper.go:62]: should we do basic validation to `authority` here? Or there is the other place to validate the keeper?

Comment (adu-crypto) [reply x/cronos/keeper/keeper.go:62]: typically this `authority` is expected to be the govern module address, but yes I think it's worth to do some basic validations.

File: x/cronos/module.go (hunk 2 above)
Comment (adu-crypto) [reply x/cronos/module.go]: done

File: x/cronos/genesis.go (hunk 1 above)
Comment (adu-crypto) [reply x/cronos/genesis.go]: done

File: proto/cronos/query.proto [hunk 4]
```diff
@@ -24,6 +25,11 @@ service Query {
     // ReplayBlock replay the eth messages in the block to recover the results of false-failed txs.
     rpc ReplayBlock(ReplayBlockRequest) returns (ReplayBlockResponse) { }
 
+    // Params queries all parameters.
+    rpc Params(QueryParamsRequest) returns (QueryParamsResponse) {
+        option (google.api.http).get = "/cronos/v1/params";
```
Comment (mmsqe) [inline proto/cronos/query.proto:30]: not sure if we'd enable `module_query_safe` here

File: x/cronos/migrations/v2/migrate_test.go [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+	suite.app = app.Setup(suite.T(), cronosAdmin, true)
+	suite.ctx = suite.app.NewContext(false, tmproto.Header{})
+	suite.params = types.DefaultParams()
+	suite.params.CronosAdmin = cronosAdmin
+	suite.params.EnableAutoDeployment = true
+}
+
+func (suite *MigrationTestSuite) TestMigrate() {
+	suite.Require().Equal(suite.params, suite.app.CronosKeeper.GetParams(suite.ctx))
```
Comment (mmsqe) [inline x/cronos/migrations/v2/migrate_test.go:46]: Will we test real migration here?

File: x/cronos/module.go [hunk 6]
```diff
@@ -114,17 +116,19 @@ func (AppModuleBasic) GetQueryCmd() *cobra.Command {
 // AppModule implements the AppModule interface for the capability module.
 type AppModule struct {
 	AppModuleBasic
-	keeper        keeper.Keeper
-	accountKeeper types.AccountKeeper
-	bankKeeper    types.BankKeeper
+	keeper         keeper.Keeper
+	accountKeeper  types.AccountKeeper
+	bankKeeper     types.BankKeeper
+	legacySubspace paramstypes.Subspace
```
Comment (mmsqe) [inline x/cronos/module.go:122]: May I ask if `legacySubspace` only used for migration?

File: x/cronos/migrations/v2/migrate_test.go (hunk 5 above)
Comment (adu-crypto) [reply x/cronos/migrations/v2/migrate_test.go:46]: yeah, this should not call `migrate` here. I need to think about how to test real migration

File: x/cronos/module.go (hunk 6 above)
Comment (adu-crypto) [reply x/cronos/module.go:122]: yes

File: x/cronos/types/messages.go [hunk 7]
```diff
@@ -171,3 +176,32 @@ func (msg *MsgUpdateTokenMapping) GetSignBytes() []byte {
...
+}
+
+// GetSigners returns the expected signers for a MsgUpdateParams message.
+func (msg *MsgUpdateParams) GetSigners() []sdk.AccAddress {
+	addr, err := sdk.AccAddressFromBech32(msg.Authority)
+	if err != nil {
+		panic(err)
+	}
+	return []sdk.AccAddress{addr}
+}
```
Comment (mmsqe) [inline x/cronos/types/messages.go:194]:
```suggestion
func getSigners(adr string) []sdk.AccAddress {
	address, err := sdk.AccAddressFromBech32(adr)
	if err != nil {
		panic(err)
	}
	return []sdk.AccAddress{address}
}

// GetSigners returns the expected signers for a MsgUpdateParams message.
func (msg *MsgUpdateParams) GetSigners() []sdk.AccAddress {
	return getSigners(msg.Authority)
}
```

Comment (mmsqe) [reply x/cronos/types/messages.go:194]: could apply for rest `GetSigners()`

File: x/cronos/types/messages.go [hunk 8]
```diff
@@ -171,3 +176,32 @@ func (msg *MsgUpdateTokenMapping) GetSignBytes() []byte {
...
+
+// ValidateBasic does a sanity check on the provided data.
+func (msg *MsgUpdateParams) ValidateBasic() error {
+	if _, err := sdk.AccAddressFromBech32(msg.Authority); err != nil {
+		return sdkerrors.Wrap(err, "invalid authority address")
+	}
+
+	if err := msg.Params.Validate(); err != nil {
```
Comment (mmsqe) [inline x/cronos/types/messages.go:202]: Do we need validate for ibc related params here?

Comment (adu-crypto) [reply x/cronos/types/messages.go:202]: yeah and `Params.Validate()` does what you want.

File: x/cronos/types/messages.go (hunk 7 above)
Comment (adu-crypto) [reply x/cronos/types/messages.go:194]: is this better design pattern?

File: x/cronos/types/messages.go (hunk 8 above)
Comment (mmsqe) [reply x/cronos/types/messages.go:202]: might just fix upgrade test then, complain `IbcCroDenom` not registered

File: x/cronos/types/messages.go (hunk 7 above)
Comment (mmsqe) [reply x/cronos/types/messages.go:194]: just simple reuse, since 4 `GetSigners` already

File: x/cronos/types/messages.go (hunk 8 above)
Comment (adu-crypto) [reply x/cronos/types/messages.go:202]: how could we check if `IbcCroDenom` is registered or not?

File: proto/cronos/query.proto (hunk 4 above)
Comment (adu-crypto) [reply proto/cronos/query.proto:30]: what's this option used for? I didn't see such options in the context.

Comment (mmsqe) [reply proto/cronos/query.proto:30]: also new, but you could trace [here](https://github.com/cosmos/cosmos-sdk/blob/main/proto/cosmos/query/v1/query.proto)

File: x/cronos/types/messages.go (hunk 8 above)
Comment (mmsqe) [reply x/cronos/types/messages.go:202]: seems we miss register for [cronos](https://github.com/adu-crypto/cronos/blob/73f9ebdaf12408e15e163a7fa591ef8c85a166b9/app/app.go#L971) 
```paramsKeeper.Subspace(cronostypes.ModuleName).WithKeyTable(cronostypes.ParamKeyTable())```

File: proto/cronos/query.proto (hunk 4 above)
Comment (adu-crypto) [reply proto/cronos/query.proto:30]: seems indeed safe for module queries by itself.
but we may upgrade this in another pr.

File: x/cronos/types/keys.go [hunk 9]
```diff
@@ -31,6 +31,9 @@ var (
 	KeyPrefixDenomToExternalContract = []byte{prefixDenomToExternalContract}
 	KeyPrefixDenomToAutoContract     = []byte{prefixDenomToAutoContract}
 	KeyPrefixContractToDenom         = []byte{prefixContractToDenom}
+
+	// ParamsKey is the key for params.
+	ParamsKey = []byte{0x00}
```
Comment (yihuang) [inline x/cronos/types/keys.go:36]: Should be added the enum with the other key prefixes.

**Review by yihuang — CHANGES_REQUESTED** [review]
Is this PR ready, can you add an integration test?

Also we store the parameters in module store, can we still reuse the `ParamSet`?

**Comment by adu-crypto** [comment]
> Is this PR ready, can you add an integration test for the gov param message?

yeah sure. I will add the integration test soon

File: x/cronos/types/keys.go (hunk 9 above)
Comment (adu-crypto) [reply x/cronos/types/keys.go]: done

File: x/cronos/migrations/v2/migrate_test.go (hunk 5 above)
Comment (adu-crypto) [reply x/cronos/migrations/v2/migrate_test.go]: testing real migration requires either two binaries with different versions or a before-upgrade database, which seems not easy to be done. Do you have better ideas?

Comment (mmsqe) [reply x/cronos/migrations/v2/migrate_test.go]: I think a [test](https://github.com/adu-crypto/cronos/pull/6/commits/b7dd95e8434fb73ee5fc29df719e746f111fb858) for migrate with store check is enough.

**Review by mmsqe — APPROVED** [review]

**Comment by yihuang** [comment]
> > Is this PR ready, can you add an integration test for the gov param message?
> 
> yeah sure. I will add the integration test soon

and ideally check the migration is succesfully in the upgrade integration test.

**Comment by adu-crypto** [comment]
@yihuang all checks has passed except protobuf breakage, how could I escape this check?

**Review by yihuang — CHANGES_REQUESTED** [review]
LGTM overall, can you do a gov proposal to update parameters in integration test?

**Comment by adu-crypto** [comment]
> LGTM overall, can you do a gov proposal to update parameters in integration test?

sure, add soon

File: integration_tests/test_gov_update_params.py [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+    wait_for_block_time,
+    parse_events,
+)
+
+
+@pytest.fixture(scope="module")
+def cronos_gov(tmp_path_factory):
+    path = tmp_path_factory.mktemp("cronos-gov")
+    yield from setup_custom_cronos(
+        path, 26300, Path(__file__).parent / "configs/default.jsonnet"
```
Comment (yihuang) [inline integration_tests/test_gov_update_params.py:26]: You can use the existing cronos fixture.

**Comment by yihuang** [comment]
can you fix the conflicts?

File: integration_tests/test_gov_update_params.py (hunk 10 above)
Comment (adu-crypto) [reply integration_tests/test_gov_update_params.py:26]: directly using `cronos` like this:
```python
def test_gov_update_params(cronos, tmp_path):
...
```

**Comment by adu-crypto** [comment]
> can you fix the conflicts?

No problem but `test_gov_update_params` failed for 

> AssertionError:  (cronosd tx gov submit-proposal /private/tmp/pytest-of-adudu/pytest-2/test_gov_update_params0/proposal.json -y --home /private/tmp/pytest-of-adudu/pytest-2/cronos-gov0/cronos_777-1/node0 --gas-prices 100000000000basetcro --gas auto --gas-adjustment 1.5 --from community)

I'm still trying to fix this

File: integration_tests/test_gov_update_params.py (hunk 10 above)
Comment (adu-crypto) [reply integration_tests/test_gov_update_params.py]: done

File: x/cronos/module.go [hunk 11]
```diff
@@ -154,6 +154,7 @@ func (am AppModule) LegacyQuerierHandler(legacyQuerierCdc *codec.LegacyAmino) sd
 // module-specific GRPC queries.
 func (am AppModule) RegisterServices(cfg module.Configurator) {
 	types.RegisterQueryServer(cfg.QueryServer(), am.keeper)
+	types.RegisterMsgServer(cfg.MsgServer(), keeper.NewMsgServerImpl(am.keeper))
```
Comment (yihuang) [inline x/cronos/module.go:157]: Interesting, it wasn't registered before, how the `update_token_mapping` msg works before?

**Review by yihuang — APPROVED** [review]
LGTM

**Pull request merged**